# 1. Introduction

We're deciding whether we should wrestle a bear that's in front of us. We're using various criteria to make our final decision, including the size of the bear, and whether escape is possible

<block><pre>
Let's say we had a data set of people who survived bear encounters and the actions they took:

Bear name    Size    Escape possible?    Action
Yogi         Small   No                  Wrestle
Winnie       Small   Yes                 Wrestle
Baloo        Large   Yes                 Run away
Gentle Ben   Large   No                  Wrestle

<block></pre>

If we wanted to optimize our chances of surviving a bear encounter, we could construct a decision tree to tell us what action to take.

As our data set gets larger though, this becomes less and less practical. What if we had 10000 rows and 10 variables? Would you want to look through all of the possibilities to construct a tree?

This is where the decision tree machine learning algorithm can help. It enables us to automatically construct a decision tree that tells us what outcomes we should predict in certain situations.

**The decision tree algorithm is a supervised learning algorithm -- we first construct the tree with historical data, and then use it to predict an outcome**. `One of the major advantages of decision trees is that they can pick up nonlinear interactions between variables in the data that linear regression can't`. In our bear wrestling example, a decision tree can pick up on the fact that you should only wrestle large bears when escape is impossible, whereas a linear regression would have had to weight both factors in the absence of the other.

# 2. Overview of the Data Set

We'll be looking at individual income in the United States. The data is from the 1994 census, and contains information on an individual's marital status, age, type of work, and more. The target column, or what we want to predict, is whether individuals make less than or equal to 50k a year, or more than 50k a year.

You can download the data from the [University of California, Irvine's website](http://archive.ics.uci.edu/ml/datasets/Adult).

In [1]:
import pandas as pd

# Set index_col to False to avoid pandas thinking that the first column is row indexes (it's age)
income = pd.read_csv("income.csv", index_col=False)
print(income.head(5))

   age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country high_income  
0          2174             0              40   United-States   

# 3. Converting Categorical Variables

As we can see in our data, we have categorical variables such as workclass that have string values. Multiple individuals can share the same string value. The types of work include State-gov, Self-emp-not-inc, Private, and so on. Each of these strings is a label for a category. Another example of a column of categories is sex, where the options are Male and Female.

`Before we get started with decision trees, we need to convert the categorical variables in our data set to numeric variables. This involves assigning a number to each category label, then converting all of the labels in a column to the corresponding numbers.`

One strategy is to convert the columns to a categorical type. Under this approach, pandas will display the labels as strings, but internally store them as numbers so we can do computations with them. The numbers aren't always compatible with other libraries like Scikit-learn, though, so it's easier to just do the conversion to numeric upfront.

## TODO:
* Convert the rest of the categorical columns in income (education, marital_status, occupation, relationship, race, sex, native_country, and high_income) to numeric columns.

In [2]:
for col_name in ['education','marital_status','occupation','relationship','race','sex','native_country','high_income','workclass']:
    income[col_name]=pd.Categorical(income[col_name]).codes

In [3]:
income.head()

age  workclass  fnlwgt  education  education_num  marital_status  \
0   39          7   77516          9             13               4   
1   50          6   83311          9             13               2   
2   38          4  215646         11              9               0   
3   53          4  234721          1              7               2   
4   28          4  338409          9             13               2   

   occupation  relationship  race  sex  capital_gain  capital_loss  \
0           1             1     4    1          2174             0   
1           4             0     4    1             0             0   
2           6             1     4    1             0             0   
3           6             0     2    1             0             0   
4          10             5     2    0             0             0   

   hours_per_week  native_country  high_income  
0              40              39            0  
1              13              39            0  
2              40              39            0  
3              40              39            0  
4              40               5            0

# 4. Splitting Data

**A decision tree is made up of a `series of nodes and branches`. A node is where we split the data based on a variable, and a branch is one side of the split.**

This is exactly how a decision tree works -- we keep splitting the data based on variables. As we do this, the tree accumulates more levels. The tree we made above is two levels deep because it has one split, and two "levels" of nodes.

# 5. Creating Splits

Now think of rows of data flowing through a decision tree. Next, we'll split the data set into two portions based on whether the individual works in the private sector or not

## TODO:
Split income into two parts based on the value of the workclass column.

* private_incomes should contain all rows where workclass is 4.
* public_incomes should contain all rows where workclass is not 4.

In [4]:
private_incomes=income[income['workclass']==4]
public_incomes=income[income['workclass']!=4]

In [5]:
private_incomes.head()

age  workclass  fnlwgt  education  education_num  marital_status  \
2   38          4  215646         11              9               0   
3   53          4  234721          1              7               2   
4   28          4  338409          9             13               2   
5   37          4  284582         12             14               2   
6   49          4  160187          6              5               3   

   occupation  relationship  race  sex  capital_gain  capital_loss  \
2           6             1     4    1             0             0   
3           6             0     2    1             0             0   
4          10             5     2    0             0             0   
5           4             5     4    0             0             0   
6           8             1     2    0             0             0   

   hours_per_week  native_country  high_income  
2              40              39            0  
3              40              39            0  
4              40               5            0  
5              40              39            0  
6              16              23            0

In [6]:
public_incomes.head(3)

age  workclass  fnlwgt  education  education_num  marital_status  \
0   39          7   77516          9             13               4   
1   50          6   83311          9             13               2   
7   52          6  209642         11              9               2   

   occupation  relationship  race  sex  capital_gain  capital_loss  \
0           1             1     4    1          2174             0   
1           4             0     4    1             0             0   
7           4             0     4    1             0             0   

   hours_per_week  native_country  high_income  
0              40              39            0  
1              13              39            0  
7              45              39            1

In [7]:
print(private_incomes.shape)
print(public_incomes.shape)

(22696, 15)
(9865, 15)


# 6. Decision Trees as Flows of Data

When we performed the split, 9865 rows went to the left, where workclass does not equal 4, and 22696 rows went to the right, where workclass equals 4.

It's useful to think of a decision tree as a flow of rows of data. When we make a split, some rows will go to the right, and some will go to the left. `As we build the tree deeper and deeper, each node will "receive" fewer and fewer rows.`

# 7. Splitting Data to Make Predictions

The nodes at the bottom of the tree, where we decide to stop splitting, are called `terminal nodes, or leaves`. When we do our splits, we aren't doing them randomly; we have an objective. `Our goal is to ensure that we can make a prediction on future data. In order to do this, all rows in each leaf must have only one value for our target column`. 

We're trying to predict the high_income column.
If high_income is 1, it means that the person has an income higher than 50k a year.
If high_income is 0, it means that they have an income less than or equal to 50k a year.

After constructing a tree using the income data, we'll want to make predictions. In order to do this, we'll take a new row and feed it through our decision tree.

First, we check whether the person works in the private sector. If they do, we'll check to see whether they're native to the US, and so on.

Eventually, we'll reach a leaf. The leaf will tell us what value we should predict for high_income.

`In order to be able to make this prediction, all of the rows in a leaf should only have a single value for high_income. This means that leaves can't have both 0 and 1 values in the high_income column. Each leaf can only have rows with the same values for our target column. If this isn't the case, we won't be able to make effective predictions.`

`We'll need to continue splitting nodes until we get to a point where all of the rows in a node have the same value for high_income.`

# 8. Overview of Data Set Entropy

Because we're trying to reach the leaves having only 1s or only 0s in high_income, each split will need to get us closer to that goal.

When we split, we'll try to separate as many 0s from 1s in the high_income column as we can. In order to do this, we need a metric for how "together" the different values in the high_income column are.

Data scientists commonly use a metric called `entropy for this purpose`. `Entropy refers to disorder. The more "mixed together" 1s and 0s are, the higher the entropy. A data set consisting entirely of 1s in the high_income column would have low entropy`.

Entropy, which is not to be confused with entropy from physics, comes from [information theory](https://en.wikipedia.org/wiki/Information_theory). Information theory is based on probability and statistics, and deals with the transmission, processing, utilization, and extraction of information. A key concept in information theory is the notion of a bit of information. One bit of information is [one unit of information](https://en.wikipedia.org/wiki/Units_of_information).

We can represent a bit of information as a binary number because it either has the value 1 or 0. Suppose there's an equal probability of tomorrow being sunny (1) or not sunny (0). If I tell you that it will be sunny, I've given you one bit of information.

Entropy can be much more complex, especially when we get to cases with more than two possible outcomes, or differential probabilities. A deep understanding of entropy isn't necessary for constructing decision trees, however. If you'd like, you can read more about entropy at [Wikipedia]().

## $-\sum_{i=1}^{c} {\mathrm{P}(x_i) \log_b \mathrm{P}(x_i)}$

We iterate through each unique value in a single column (in this case, high_income), and assign it to $i$. We then compute the probability of that value occurring in the data ($\mathrm{P}(x_i)$). Next we do some multiplication, and sum all of the values together. $b$ is the base of the logarithm. We commonly use the value 2 for this, but we can also set it to 10 or another value.

In [8]:
'''Let's say we have this data:

age    high_income
25     1
50     1
30     0
50     0
80     1
'''

"Let's say we have this data:\n\nage    high_income\n25     1\n50     1\n30     0\n50     0\n80     1\n"

$\begin{align}
-\sum_{i=1}^{c} {\mathrm{P}(x_i) \log_b \mathrm{P}(x_i)} \nonumber \\
-((2/5 * \log_2 2/5) + (3/5 * \log_2 3/5)) \nonumber \\ 
-(-0.5287712379549449 + -0.44217935649972373)\nonumber \\
.97 \nonumber \\
\end{align}$

We get less than one "bit" of information -- only .97 -- because there are slightly more 1s in the sample data than 0s. This means that if we were predicting a new value, we could guess that the answer is 1 and be right more often than wrong (because there's a .6 probability of the answer being 1). Due to this prior knowledge, we gain less than a full "bit" of information when we observe a new value.

# 9. Overview of Data Set Entropy

## TODO:
*Compute the entropy of the high_income column in the income dataframe, and assign the result to income_entropy.

In [9]:
import math

prob_0=(income[income['high_income']==0].shape[0])/income['high_income'].shape[0]
prob_1=(income[income['high_income']==1].shape[0])/income['high_income'].shape[0]


In [10]:
income_entropy=-(prob_0*math.log(prob_0,2) + prob_1*math.log(prob_1,2))

In [11]:
income_entropy

0.7963839552022132

# 10. Information Gain

**`We'll need a way to go from computing entropy to figuring out which variable to split on. We can do this using information gain, which tells us which split will reduce entropy the most.`**

### Here's the formula for information gain:
## $IG(T,A) = Entropy(T)-\sum_{v\in A}\frac{|T_{v}|}{|T|} \cdot Entropy(T_{v})$

**We're computing information gain ($IG$) for a given target variable ($T$), as well as a given variable we want to split on ($A$).**

To compute it, we first calculate the entropy for $T$. Then, for each unique value $v$ in the variable $A$, we compute the number of rows in which $A$ takes on the value $v$, and divide it by the total number of rows. Next, we multiply the results by the entropy of the rows where  is . We add all of these subset entropies together, then subtract from the overall entropy to get information gain.

Here's an alternate explanation. `We're finding the entropy of each set post-split, weighting it by the number of items in each split, then subtracting from the current entropy. If the result is positive, we've lowered entropy with our split. The higher the result is, the more we've lowered entropy.`

`One strategy for constructing trees is to create as many branches at each node as there are unique values for the variable we're splitting on. So if the variable has three or four values, we'd end up with three or four branches. This approach usually involves more complexity than it's worth and doesn't improve prediction accuracy, but it's worth knowing about.`

To simplify the calculation of information gain and make splits simpler, we won't do it for each unique value. We'll find the median for the variable we're splitting on instead. Any rows where the value of the variable is below the median will go to the left branch, and the rest of the rows will go to the right branch. To compute information gain, we'll only have to compute entropies for two subsets.

# 11. Information Gain

## TODO:
* Compute the information gain for splitting on the age column of income.

* First, compute the median of age.
* Then, assign anything less than or equal to the median to the left branch, and anything greater than the median to the right branch.
* Compute the information gain and assign it to age_information_gain.

In [12]:
import numpy

def calc_entropy(column):
    """
    Calculate entropy given a pandas series, list, or numpy array.
    """
    # Compute the counts of each unique value in the column
    counts = numpy.bincount(column)
    # Divide by the total column length to get a probability
    probabilities = counts / len(column)
    
    # Initialize the entropy to 0
    entropy = 0
    # Loop through the probabilities, and add each one to the total entropy
    for prob in probabilities:
        if prob > 0:
            entropy += prob * math.log(prob, 2)
    
    return -entropy

# Verify that our function matches our answer from earlier
entropy = calc_entropy([1,1,0,0,1])
print(entropy)

information_gain = entropy - ((.8 * calc_entropy([1,1,0,0])) + (.2 * calc_entropy([1])))
print(information_gain)
income_entropy = calc_entropy(income["high_income"])

median_age = income["age"].median()

left_split = income[income["age"] <= median_age]
right_split = income[income["age"] > median_age]

age_information_gain = income_entropy - ((left_split.shape[0] / income.shape[0]) * calc_entropy(left_split["high_income"]) + ((right_split.shape[0] / income.shape[0]) * calc_entropy(right_split["high_income"])))

0.9709505944546686
0.17095059445466854


# 12. Finding the Best Split

## TODO:
* Create a list called information_gains.
  * It should contain, in order, the information gain from splitting on these columns: age, workclass, education_num, marital_status, occupation, relationship, race, sex, hours_per_week, native_country.
* Find the highest value in the information_gains list, and assign the corresponding column name to highest_gain.

In [13]:
def calc_information_gain(data, split_name, target_name):
    """
    Calculate information gain given a data set, column to split on, and target
    """
    # Calculate the original entropy
    original_entropy = calc_entropy(data[target_name])
    
    # Find the median of the column we're splitting
    column = data[split_name]
    median = column.median()
    
    # Make two subsets of the data, based on the median
    left_split = data[column <= median]
    right_split = data[column > median]
    
    # Loop through the splits and calculate the subset entropies
    to_subtract = 0
    for subset in [left_split, right_split]:
        prob = (subset.shape[0] / data.shape[0]) 
        to_subtract += prob * calc_entropy(subset[target_name])
    
    # Return information gain
    return original_entropy - to_subtract

# Verify that our answer is the same as on the last screen
print(calc_information_gain(income, "age", "high_income"))

columns = ["age", "workclass", "education_num", "marital_status", "occupation", "relationship", "race", "sex", "hours_per_week", "native_country"]
information_gains = []
# Loop through and compute information gains
for col in columns:
    information_gain = calc_information_gain(income, col, "high_income")
    information_gains.append(information_gain)

# Find the name of the column with the highest gain
highest_gain_index = information_gains.index(max(information_gains))
highest_gain = columns[highest_gain_index]

0.047028661304691965



# 13. Build the Whole Tree

 To construct an entire tree, we'll need to continue creating splits until the leaves only have a single class.

**So far, we've been following the [`ID3 algorithm`](https://en.wikipedia.org/wiki/ID3_algorithm) to construct decision trees. There are other algorithms like [`CART`](https://www.datasciencecentral.com/profiles/blogs/introduction-to-classification-regression-trees-cart) that use different measures for the split criterion.**
